lib/PrintAsClang/DeclAndTypePrinter.cpp

```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
+    printFunctionClangAttributes(FD, funcTy);
+    printAvailability(FD);
     os << ";\n";
   }
 
+  struct FuncionSwiftABIInformation {
+    FuncionSwiftABIInformation(FuncDecl *FD, Mangle::ASTMangler &mangler) {
+      isCDecl = FD->getAttrs().hasAttribute<CDeclAttr>();
+      if (!isCDecl) {
+        mangledName = mangler.mangleAnyDecl(FD, /*prefix=*/true);
+      } else {
+        cdeclName = FD->getAttrs().getAttribute<CDeclAttr>()->Name;
+      }
+    }
+
+    StringRef getSymbolName() const {
+      return isCDecl ? cdeclName : mangledName;
+    }
+
+    bool useCCallingConvention() const { return isCDecl; }
+
+    bool useMangledSymbolName() const { return !isCDecl; }
+
+  private:
+    bool isCDecl;
+    std::string mangledName;
+    StringRef cdeclName;
+  };
+
+  FuncionSwiftABIInformation
+  printSwiftABIFunctionSignatureAsCxxFunction(FuncDecl *FD) {
+    assert(outputLang == OutputLanguageMode::Cxx);
+    Optional<ForeignAsyncConvention> asyncConvention =
+        FD->getForeignAsyncConvention();
+    Optional<ForeignErrorConvention> errorConvention =
+        FD->getForeignErrorConvention();
+    assert(!FD->getGenericSignature() &&
+           "top-level generic functions not supported here");
+    // FIXME (Alex): Make type adjustments for C++.
+    auto funcTy = FD->getInterfaceType()->castTo<FunctionType>();
+    auto resultTy =
+        getForeignResultType(FD, funcTy, asyncConvention, errorConvention);
+
+    Mangle::ASTMangler mangler;
+    FuncionSwiftABIInformation funcABI(FD, mangler);
+
+    os << "extern \"C\" ";
+    printFunctionDeclAsCFunctionDecl(FD, funcABI.getSymbolName(), resultTy);
+    // Swift functions can't throw exceptions, we can only
+    // throw them from C++ when emitting C++ inline thunks for the Swift
+    // functions.
+    os << " noexcept";
+    if (!funcABI.useCCallingConvention())
+      os << " SWIFT_CALL";
+    printAvailability(FD);
+    os << ';';
+    if (funcABI.useMangledSymbolName()) {
+      // add a comment with a demangled function name.
+      os << " // ";
+      FD->getName().print(os);
+    }
+    os << "\n";
+    return funcABI;
+  }
+
+  void printAbstractFunctionAsCxxFunctionThunk(
+      FuncDecl *FD, const FuncionSwiftABIInformation &funcABI) {
+    printDocumentationComment(FD);
+
+    Optional<ForeignAsyncConvention> asyncConvention =
+        FD->getForeignAsyncConvention();
+    Optional<ForeignErrorConvention> errorConvention =
+        FD->getForeignErrorConvention();
+    assert(!FD->getGenericSignature() &&
+           "top-level generic functions not supported here");
+    auto funcTy = FD->getInterfaceType()->castTo<FunctionType>();
+    auto resultTy =
+        getForeignResultType(FD, funcTy, asyncConvention, errorConvention);
+
+    os << "inline ";
+    printFunctionDeclAsCFunctionDecl(FD,
+                                     FD->getName().getBaseIdentifier().get(),
+                                     resultTy, /*printEmptyParamNames=*/true);
```

Comment:
Why only print empty param names some places? Seems like that should maybe be a global config. Then you also wouldn't have to launder an int down to `print`.

Reply:
we need to print out empty parameters for the inline thunks specifically as we need to use them in the body.